sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp

```diff
@@ -417,6 +417,57 @@ namespace Azure { namespace Core { namespace Http { namespace Policies {
           Context const& context) const override;
     };
 
+    /**
+     * @brief Helper class for creating scopes for the token request context.
+     *
+     */
+    struct TokenScopes
+    {
+      /**
+       * @brief Parses url and generates a scope from it.
+       *
+       * @details For example, from a url like: `https://<account>.<service>.azure.net` the
+       * generated scope is in the form of: `https://<service>.azure.net/.default`.
+       *
+       * @remark As per Azure rules, the account for any service  name must only contain
+       * alphanumeric characters and dashes and cannot start with a number. This allow this method
+       * to calculate the scope after the first `.` from the url host.
+       *
+       * @param url The url used to create an Azure SDK client.
+       * @param defaultScope Optional hardcoded scope to be always added to the result scopes. It
```

Comment:
Why would you need `https://vault.azure.net/.default` to also be added, it does not work without it?

Reply:
This is seeking for backwards compatibility more than anything.

The current GA versions are hardcoding a scope, and some customers might have already be expecting that to happen regardless the provided url.  If for any reason this change, some customers might be broken.

This fix of leaving the current one is only due to that.  Clients can continue adding the old scope regardless that's in the url, plus the new scope.

Reply:
But the provided URL is going to be supplied as something like `https://vhvb-kv.vault.azure.net`, and this function will drop the `vhvb-kb`, making the scope identical to the hardcoded one, `https://vault.azure.net/.default`, wouldn't it?

Reply:
Yes, in case it is identical, the default scope is ignored.  But I wanted to add this to prevent any non-considered situation where the current behavior changes due to the change.

Reply:
But you are going to end up requestion the access for scope, where the user not necessarily have access. We should not be requesting more than we need. Or I am not getting it.

Reply:
it's an optional parameter.  It's invisible to the user, just like right now it is.

Right now, when creating a client, we are hardcoding the scope.  After this change, the scope will be calculated, and there's a change for clients to keep adding the current hardcoded string so the change is only an addition to current behavior.

If a service wants to avoid adding a default scope, it's fine. For instance, new services, like Key Vault Secrets can't use the calculated one only since the beginning.

Only GA packages need to decide if they want to remove the default scope.

Reply:
If a key vault's URL is `https://mykv.vault.privatecloud.net`, we should only be requesting `https://vault.privatecloud.net/.default` scope. We should not be requesting a token for `https://vault.azure.net.net/.default` scope.

Reply:
Here's what .NET does: https://github.com/Azure/azure-sdk-for-net/blob/4162f6fa2445b2127468b9cfd080f01c9da88eba/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L139-L152
They do not request the token for both scopes.
BTW, they do keep the port, unless it is 0.